extensions/amp-viewer-assistance/0.1/amp-viewer-assistance.js

```diff
@@ -173,19 +176,49 @@ export class AmpViewerAssistance {
       return false;
     }
     const update = args['update'];
+    const error = args['error'];
     const actionStatus = update && update['actionStatus'];
-    if (!update || !actionStatus) {
+
+    if (error) {
+      return true;
+    } else if (update) {
+      if (!actionStatus || !ACTION_STATUS_WHITELIST.includes(actionStatus)) {
+        user().error(TAG, '"updateActionState" action "update" parameter' +
+            ' must contain a valid "actionStatus" field.');
+        return false;
+      }
+      return true;
+    } else {
       user().error(TAG, '"updateActionState" action must have an' +
-          ' an "update" parameter containing an "actionStatus" field.');
+          ' "update" or "error" parameter.');
       return false;
     }
-    if (!ACTION_STATUS_WHITELIST.includes(actionStatus)) {
-      user().error(TAG, 'Invalid "update.actionStatus" value for ' +
-          '"updateActionState":', actionStatus);
-      return false;
+  }
+
+  /**
+   * Transforms an error response object into a suitable updateActionState
+   * payload to send to the viewer.
+   * @private
+   * @param {?JsonObject} args
+   * @return {!Promise<!JsonObject>}
+   */
+  transformUpdateActionStateArgs_(args) {
+    if (args['update']) {
+      return Promise.resolve(args);
     }
-    user().info(TAG, 'Sending "actionStatus":', actionStatus);
-    return true;
+    // Must transform 'error' Response object
+    const response = args['error'];
+    return response.text().then(errorMessage => {
```

Comment:
Might be worth handling the case where `args.error` is not of type `Response`:

```js
const response = args['error'];
if (response && typeof response.text === 'function') {
  ...
}
```
Or maybe check this in `isValidActionStatusArgs_()` (or combine the functions so the checks are next to their usage).

Reply:
Right, makes sense. Combined the two functions into a single function. It rejects the promise on an invalid input which passes the failed message to the catch block:
```js
}).catch(err => {
  user().error(TAG, err.toString());
});
```